ibis/backends/mysql/__init__.py

```diff
@@ -121,6 +124,215 @@ def begin(self):
+                    typ = dt.int16
+                elif field_length <= 32:
+                    typ = dt.int32
+                elif field_length <= 64:
+                    typ = dt.int64
+                else:
+                    assert False, "invalid field length for BIT type"
+            else:
+                if flags.is_set:
+                    # sets are limited to strings
+                    typ = dt.Set(dt.string)
+                elif flags.is_unsigned and flags.is_num:
+                    typ = getattr(dt, f"U{typ.__name__}")
+                elif type_code in TEXT_TYPES:
+                    # binary text
+                    if field.charsetnr == MY_CHARSET_BIN:
+                        typ = dt.Binary
+                    else:
+                        typ = dt.String
+
+            # projection columns are always nullable
+            dtype = typ(nullable=True)
+            fields.append((name, dtype))
+
+        return sch.Schema.from_tuples(fields)
+
+    def _get_temp_view_definition(
+        self,
+        name: str,
+        definition: sa.sql.compiler.Compiled,
+    ) -> str:
+        return f"CREATE OR REPLACE VIEW {name} AS {definition}"
+
+    def _register_temp_view_cleanup(self, name: str) -> None:
+        atexit.register(self.con.execute, f"DROP VIEW IF EXISTS {name}")
+
+
+# ported from my_decimal.h:my_decimal_length_to_precision in mariadb
+def _decimal_length_to_precision(
+    *,
+    length: int,
+    scale: int,
+    is_unsigned: bool,
+) -> int:
+    return length - (scale > 0) - (not (is_unsigned or not length))
+
+
+_type_codes = {
+    0: "DECIMAL",
+    1: "TINY",
+    2: "SHORT",
+    3: "LONG",
+    4: "FLOAT",
+    5: "DOUBLE",
+    6: "NULL",
+    7: "TIMESTAMP",
+    8: "LONGLONG",
+    9: "INT24",
+    10: "DATE",
+    11: "TIME",
+    12: "DATETIME",
+    13: "YEAR",
+    15: "VARCHAR",
+    16: "BIT",
+    245: "JSON",
+    246: "NEWDECIMAL",
+    247: "ENUM",
+    248: "SET",
+    249: "TINY_BLOB",
+    250: "MEDIUM_BLOB",
+    251: "LONG_BLOB",
+    252: "BLOB",
+    253: "VAR_STRING",
+    254: "STRING",
+    255: "GEOMETRY",
+}
+
+
+_type_mapping = {
+    "DECIMAL": dt.Decimal,
+    "TINY": dt.Int8,
+    "SHORT": dt.Int16,
+    "LONG": dt.Int32,
+    "FLOAT": dt.Float32,
+    "DOUBLE": dt.Float64,
+    "NULL": dt.Null,
+    "TIMESTAMP": lambda nullable: dt.Timestamp(
+        timezone="UTC",
+        nullable=nullable,
+    ),
+    "LONGLONG": dt.Int64,
+    "INT24": dt.Int32,
+    "DATE": dt.Date,
+    "TIME": dt.Time,
+    "DATETIME": dt.Timestamp,
+    "YEAR": dt.Int16,
+    "VARCHAR": dt.String,
+    "BIT": dt.Int8,
+    "JSON": dt.JSON,
+    "NEWDECIMAL": dt.Decimal,
+    "ENUM": dt.String,
+    "SET": lambda nullable: dt.Set(dt.string, nullable=nullable),
+    "TINY_BLOB": dt.Binary,
+    "MEDIUM_BLOB": dt.Binary,
+    "LONG_BLOB": dt.Binary,
+    "BLOB": dt.Binary,
+    "VAR_STRING": dt.String,
+    "STRING": dt.String,
+    "GEOMETRY": dt.Geometry,
+}
+
+
+class _FieldFlags:
```

Comment:
Can you move some of this code to `mysql/datatypes.py` by encapsulating the type name and field flags into a `MySQLDataType` or something similar to [ClickhouseDataType](https://github.com/ibis-project/ibis/blob/master/ibis/backends/clickhouse/client.py#L40) (which eventually should be moved to its own file as well)?

Then we could register `dt.dtype(MySQLDataType)` for converting from MySQL to ibis.

Reply:
I can move this into `mysql/datatypes.py`, but I'd rather avoid doing a big refactoring/restructuring of this code for now. I'm happy to revisit once we see more use of this feature in the wild.